### ISSUE TITLE
Write linter to discourage usage of previous binding in :or defaults

### ISSUE THREAD

**Issue by borkdude** [clj-kondo/clj-kondo#916]
Update from #673:

Using a previous arg in `:or` is fine, but not if the default depends on a value in the defaults map itself.

So `(fn [{:keys [x y] :or {y x}}])` is fine, but not `(fn [{:keys [x y z] :or {y x z y}}])`, since maps are unordered and it's undefined to what `z` will be bound: either the binding `y` passed as is, or `x` if `y` was `nil`.

So clj-kondo will support the example provided by @marcomorain in #782 but will discourage `(fn [{:keys [x y z] :or {y x z y}}])` as part of this issue.

A variant of this can be found in https://github.com/borkdude/clj-kondo/issues/1011.

cc @puredanger 

See #782 and #637.

**Comment by borkdude** [comment]
@puredanger

I do have one more question about this:

```
user=> ((fn [a {:keys [:b] :or {b a}}] [a b]) 1 nil)
[1 1]
user=> ((fn [a {:keys [:b] :or {b a}}] [a b]) 1 {:b 2})
[1 2]
```

Isn't this effectively the same as:

```
(let [a 1 {:keys [:b] :or {b a}} nil] [a b]) ;;=> [1 1]
(let [a 1 {:keys [:b] :or {b a}} {:b 2}] [a b]) ;;=> [1 2]
```

Using this pattern in let is accepted right? The `fn` macro expands in a `let`:

```
user=> (macroexpand '(fn [a {:keys [:b] :or {b a}}] [a b]))
(fn* ([a p__154] (clojure.core/let [{:keys [:b], :or {b a}} p__154] [a b])))
```

So I wonder what the reason is to not accept this as officially supported. I might be overseeing something.

**Comment by puredanger** [comment]
Reading this again, I think this example is ok. We've drifted from the original thing at the start of #782 which does not seem valid. But here fn binds left to right, and let binds in order, so I think either of these should be ok and I retract what I said earlier.

**Comment by borkdude** [comment]
@puredanger 

Do you mean:

``` 
(fn unresolved-symbol-x
  [x & {:keys [y]
          :or {y (inc x)}}]
  (+ y x))
``` 

at the start of #782? How is that example any different except that it uses varargs?

It expands into:

``` 
(fn* unresolved-symbol-x ([x & p__138] (clojure.core/let [{:keys [y], :or {y (inc x)}} p__138] (+ y x))))
```

**Comment by puredanger** [comment]
Sorry, I meant #673

**Comment by seancorfield** [comment]
Another motivating example that can be used as a test case when this is implemented:

```clojure
user=> ((fn [{:keys [a b c] :foo/keys [d e f] :or {d (inc a)}}] [a b c d e f]) {:a 42 :foo/f 13})
[42 nil nil 43 nil 13]
user=> ((fn [{:keys [a b c] :foo/keys [d e f] :or {b (inc f)}}] [a b c d e f]) {:a 42 :foo/f 13})
Syntax error compiling at (REPL:1:57).
Unable to resolve symbol: f in this context
user=>
```

Although the first case "works" in this limited context, it is still undefined behavior and the "almost identical" second small example fails "as expected".

**Comment by borkdude** [comment]
@seancorfield OK, I think we're going from:

https://github.com/clj-kondo/clj-kondo/issues/782 (not supported in clj-kondo) 

to: supported in clj-kondo (it correctly recognizes the bindings)

to: clj-kondo actively warns against doing this.

Can you drop some quotes or screenshots from what Alex said here too, so we can reference this in the future?

**Comment by borkdude** [comment]
Here they are:

[2h](https://clojurians.slack.com/archives/C03S1KBA2/p1665338446370599?thread_ts=1665334951.512419&cid=C03S1KBA2)
Alex Miller (Clojure team)
  The values of :or default maps should not be expressions that depend on the locals you are binding, anything else is undefined and it’s behavior may change in the future
[2h](https://clojurians.slack.com/archives/C03S1KBA2/p1665338509289249?thread_ts=1665334951.512419&cid=C03S1KBA2)
Alex Miller (Clojure team)
  Any appearance that this works now is implementation details leaking, not intentional behavior
[1h](https://clojurians.slack.com/archives/C03S1KBA2/p1665340477352879?thread_ts=1665334951.512419&cid=C03S1KBA2)
Alex Miller (Clojure team)
  Symbol reference is fine but you should not consider the symbols being bound in the destructuring to be in scope yet. And I wouldn’t consider it wrong for Clojure to change the map type of that :or map to a built-in map type in the service of optimization.
...
Alex Miller (Clojure team)
  
[@seancorfield](https://clojurians.slack.com/team/U04V70XH6)
 if you have enough kv pairs in that map binding example earlier, the order of evaluation for those key sets may change. In short, the expression side of an :or map should not rely on locals being bound in the same destructuring map and you should not make assumptions about when those or expressions are evaluated (or even whether they are evaluated at all)

**Comment by borkdude** [comment]
Considering the upvotes, I'm going to have a stab at this soon. 

tl;dr, after reading all the comments again: "the expression side of an :or map should not rely on locals being bound in the same destructuring map"

**Comment by borkdude** [comment]
Proposed name: `:destructured-or-refers-binding-of-same-map` until someone suggest something more brief/better.

**Comment by borkdude** [comment]
I also found another fun edge case: `(let [[{:keys [b a] :or {b (inc a)}}] {}])`

**Comment by borkdude** [comment]
While implementing the linter I found this case:  https://github.com/clojure/clojurescript/blob/47a7bfdefd156027b7b98cb483ddd93add71fbb2/src/main/clojure/cljs/repl.cljc#L1051
Not only does this example use a binding in :or from the same destructuring map, the value is even more undeterministic since it depends on a default in the same :or map
